tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -130,15 +132,29 @@ public void ActivityStopped<T>(string sourceName, T activity)
                         return;
                     }
 
-                    if (ActivityMappingById.TryRemove(activity.Id, out ActivityMapping value) && value.Scope?.Span is not null)
+                    // first we look for the normal ID, then the parent
+                    if (ActivityMappingById.TryRemove(activity.Id, out ActivityMapping someValue) && someValue.Scope?.Span is not null)
+                    {
+                        // We have the exact scope associated with the Activity
+                        if (Log.IsEnabled(LogEventLevel.Debug))
+                        {
+                            Log.Debug("DefaultActivityHandler.ActivityStopped: [Source={SourceName}, Id={Id}, RootId={RootId}, OperationName={OperationName}, StartTimeUtc={StartTimeUtc}, Duration={Duration}]", new object[] { sourceName, activity.Id, activity.RootId, activity.OperationName!, activity.StartTimeUtc, activity.Duration });
+                        }
+
+                        CloseActivityScope(sourceName, activity, someValue.Scope);
+                        return;
+                    }
+
+                    // when we have an ActivityContext provided the ID used is the ParentID
+                    if (activity.ParentId is { } parentId && ActivityMappingById.TryRemove(parentId, out ActivityMapping parentValue) && parentValue.Scope?.Span is not null)
```

Comment:
Do we ever hit this branch? Maybe this is outdated now with your recent code changes. If an Activity reaches this point then we should have added it to the `ActivityMappingById` dictionary and returned through the previous block

Reply:
Not anymore, forgot to remove it